### ISSUE TITLE
[TeX] allow disabling of formatID

### ISSUE THREAD

**Issue by pkra** [mathjax/mathjax-v3#209]
Cf. the discussion at https://github.com/mathjax/MathJax/issues/1899

The formatID function is quite simple:

https://github.com/mathjax/mathjax-v3/blob/1217503c2fd89671041bc363f3ff9fb6c69725c7/mathjax3-ts/input/tex/Tags.ts#L355

This can lead to invalid IDs (and easily to duplicate IDs when working server-side).

It would be useful if one could disable formatID, e.g.,, check if its result is a string:

https://github.com/mathjax/mathjax-v3/blob/1217503c2fd89671041bc363f3ff9fb6c69725c7/mathjax2/legacy/extensions/TeX/AMSmath.js#L280

(The same technically applies to v2.)

**Comment by dpvc** [comment]
> e.g.,, check if its result is a string:

I'm not sure quite what you mean.  What result are you checking for a string? Both the `id` being passed to `formatId()` and the result from it are always strings.  I'm afraid I don't understand what test you want to be made.

Note that your second reference is to the legacy v2 TeX input jax, which is no longer being used by v3.  In the v3 case, the `formatID()` function is a configuration parameter that you can completely replace, so you should be able to do whatever you want with the label that it receives.

It looks like the `format*` configuration options didn't take it through to the v3 TeX configuration object. @zorkow, can you see how hard it would be to make that happen?

**Comment by pkra** [comment]
> Note that your second reference is to the legacy v2 TeX input jax,

Ah, sorry, I missed that. I came to this via code using v2.

> both the id being passed to formatId() and the result from it are always strings.

If (in v2) I configure `formatId: () => {}` then it won't return a string but undefined which ends up being stringified; that technically gives me valid IDs but is not ideal either.

For our purposes, we actually don't need customization here since we provide IDs ourselves - so we'd be happy with just an option to disable it. The current function easily creates invalid IDs, e.g., `\tag{$x$}` creates an invalid XML ID (which is still required in SVG output).

I should've filed this on v2 but it seemed more pragmatic to address this in v3.

**Comment by pkra** [comment]
>  since we provide IDs ourselves

To clarify: via `\cssId{}`.

**Comment by dpvc** [comment]
> If (in v2) I configure `formatId: () => {}` then it won't return a string

Ah, so you meant to check if it *didn't* return a string, and then not use it?  I'm not sure how this is supposed to work, however, as that would not provide anything for `\ref` and` \eqref` to use to set up their links to the tagged equation.

I'm afraid I'm confused by what you are trying to accomplish.  Can you give an explicit example of your use of `\cssId{}`?

Note that if the equation has a `\label`, it is the label, not the tag itself, that is used as the ID (and the line you linked to above is the one that handles that; the use of the tags is in the other two places where `formatID()` is used).  So you could use `\label` to override the tags.

Or you could use

```
n = 0;
mjAPI.config({
  MathJax: {
    TeX: {
      equationNumbers: {
        formatID: () => 'mjx-eqn-' + n++
      }
    }
  }
});
```

to get sequentially numbered labels regardless of the tag.  Or

```
n = 0;
mjAPI.config({
  MathJax: {
    TeX: {
      equationNumbers: {
        formatID: (id) => {
          const match = id.match(/\\cssId{(.*?)}/);
          return 'mjx-eq-' + (match ? match[1] : n++);
        }
      }
    }
  }
});
```

to extract the `\cssId{}` and use that for (part of) the tag's id.  Or you could define your own macro, e.g., `\tagID{}` that simply ignores its argument (so when it is typeset it does nothing), but gets extracted by the `formatID()` command to use as the (complete) tag ID, though using `\label{}` would really be just as easy.

These approaches will allow `\ref` and `\eqref` to be able to link to the tagged equations, while still giving you control over the tags used.

Does something like that fill the bill?

**Comment by pkra** [comment]
>  I'm not sure how this is supposed to work, however, as that would not provide anything for \ref and \eqref to use to set up their links to the tagged equation.

We don't need \ref and \eqref; we resolve IDs and links to them long before MathJax comes into play.

> I'm afraid I'm confused by what you are trying to accomplish. 

I want to disable formatID.

> Or you could [...] get sequentially numbered labels regardless of the tag. 

I'm aware of that but it means I have more IDs to worry about.

Anyway, I guess it's good that this issue has identified that formatID is not yet configurable in v3. 

Otherwise, this issue can be closed.

**Comment by dpvc** [comment]
> I want to disable formatID.

Saying the same thing you said originally doesn't really clarify this for me.  I don't know what "disable formatID" means.  Are you saying that you want a way not to have ids on the `<mtd>` elements that are the labels in the `<mlabeledtr>` rows?

You seem to be asking for a configuration parameter to be added to the TeX block that would prevent those ids from being added to the `<mtd>` elements.  As I mention above, that would break `\ref` and `\eqref`, and while you may not be using these yourself, the configuration parameter would not just be there for you, but for everyone, and it would cause those macros to malfunction.  That seems like a bad idea, especially when the parameter is for a pretty limited and non-standard use case.

> I have more IDs to worry about

How about using `formatID: () => ''` and having a TeX input jax post-filter that removes the empty ids on mtd elements in mlabeledtr rows?

**Comment by zorkow** [comment]
This answer is for v3 only:
I believe that this should not be configurable via an option. 
In v3 there is are tags objects with clearly defined interfaces. 
In other words, developers who want to change the behaviour

**Comment by zorkow** [comment]
Just for the record: [The definition of the interface](https://github.com/mathjax/mathjax-v3/blob/1217503c2fd89671041bc363f3ff9fb6c69725c7/mathjax3-ts/input/tex/Tags.ts#L77)

**Comment by dpvc** [comment]
@zorkow,

> developers who want to change the behaviour can subclass and add a new object to the configurations.

Can you give an example of how this would be done in practice?

What worries me is that the tag factory is not accessible to the page author (it is down in a protected property of the TeX input jax) without also having to subclass the input jax itself.  That seems like a lot of overhead to change something that was originally intended as a page-configurable function.

**Comment by zorkow** [comment]
It is accessible via the `Configuration` classes as well as via the `Options`. 
But you are probably right that for convenience we might want to supply some sort of "RawTags" subclass, which we can then make selectable via some user facing option.

**Comment by dpvc** [comment]
One last follow-up on this.  Here is a v2 extension that prevents the tag `<mtd>` elements from having an `id` attribute.  Call it `notagids.js` and put it in a folder and map the `[extensions]` path to that folder in mathjax-node, and load the extension `[extensions]/notagids.js'.

```
MathJax.Hub.Register.StartupHook("TeX AMSmath Ready", function () {
    const STACKITEM = MathJax.InputJax.TeX.Stack.Item;
    const GETTAG = STACKITEM.prototype.getTag;
    STACKITEM.Augment({
        getTag() {
            const tag = GETTAG.call(this);
            delete tag.id;
            return tag;
        }
    });
});

MathJax.Ajax.loadComplete("[extensions]/notagids.js");
```

**Comment by dpvc** [comment]
The v3 version of the extension in the previous post would be to include

```
import {AbstractTags, TagsFactory} from '../mathjax3/input/tex/Tags.js';

class NoIdTags extends AbstractTags {
    makeTag() {
        const tag = super.makeTag();
        const attributes = tag.attributes.getAllAttributes();
        delete attributes.id;
        return tag;
    }
}

TagsFactory.add('noID', NoIdTags);
```

and then use `new TeX({tags: 'noID'})` when creating the TeX input jax.

It is also possible to use something like

```
import {AbstractTags, TagsFactory} from '../mathjax3/input/tex/Tags.js';

class CustomTags extends AbstractTags {
    formatId(id) {
        return 'abc-' + id;
    }
}

TagsFactory.add('custom', CustomTags);
```

with `new TeX({tags: 'custom'})` to use a modified `formatId()` function (and similarly for the other format functions).

**Comment by zorkow** [comment]
Or you could extend one of the existing Tags classes, such as `AllTags` or `AmsTags` if you want to keep their tagging behaviour and just override the `formatId` method.
